### PR TITLE
Implement `vec_type_finalise()` in C

### PR DESCRIPTION
**Pull request by lionel-**
Also extracts `vec_recurse()` and `df_map()` as useful C utils for common vctrs patterns.

Partial frames are kept in the R side until we solidify the partial type API.

```r
x <- vctrs::vec_type(data.frame(a = 1, b = NA))

before <- vctrs::vec_type_finalise
after <- vctrs2::vec_type_finalise

bench::mark(
  before(x),
  after(x)
)[1:6]
#>   expression      min     mean   median      max `itr/sec`
#>   <chr>      <bch:tm> <bch:tm> <bch:tm> <bch:tm>     <dbl>
#> 1 before(x)    23.6µs  29.56µs   27.9µs  150.7µs    33830.
#> 2 after(x)      785ns   1.04µs    950ns   39.3µs   960796.
```

### CODE REVIEW

File: src/type.c [hunk 1]
```diff
@@ -72,6 +48,31 @@ SEXP vec_type(SEXP x) {
   return vec_type_rec(x, true);
 }
 
+
+SEXP vec_type_finalise_rec(SEXP x, bool dispatch) {
+  if (OBJECT(x)) {
```
Comment (hadley) [inline src/type.c]: I don't quite understand the logic between how you've broken up the cases here. I would expect to see `unspecified`, `data.frame`, and `partial` classes more prominently

File: src/utils.c [hunk 2]
```diff
@@ -92,6 +92,33 @@ SEXP vctrs_dispatch3(SEXP fn_sym, SEXP fn,
...
+  SEXP nms = PROTECT(Rf_getAttrib(df, R_NamesSymbol));
+  Rf_setAttrib(out, R_NamesSymbol, nms);
+
+  out = df_restore(out, df, vctrs_shared_empty_int);
+
+  UNPROTECT(2);
+  return out;
+}
+
+SEXP vec_recurse(SEXP x, SEXP (*rec)(SEXP, bool), SEXP i) {
```
Comment (hadley) [inline src/utils.c]: Why is this called `vec_recurse()`?

Comment (lionel-) [reply src/utils.c]: Because it recurses into `rec` with `dispatch = false`.

File: src/type.c (hunk 1 above)
Comment (lionel-) [reply src/type.c]: `unspecified` and `partial` are not part of the vctrs enum so are/will be handled in the `if (OBJECT(x))` branch. Partial types are kept in R for now but will be included here in the future.

File: src/utils.c (hunk 2 above)
Comment (hadley) [reply src/utils.c]: What do you mean by recurses into? Where is the recursion?

Comment (lionel-) [reply src/utils.c]: I mean `vec_method(dispatch = true) => vec_recurse() => vec_method(dispatch = false)`

Comment (hadley) [reply src/utils.c]: Yeah, I just don't find it to be a particularly evocative name

Comment (lionel-) [reply src/utils.c]: maybe `vec_dispatch()`?

**Pull request merged**